tests/unit/rankers/test_algorithms.py

```diff
@@ -149,54 +150,61 @@ def test_block_item_using_blocklist_param(self):
-            '999': MetricsModel(
-                id='home/999',
-                trailing_1_day_opens=0,
-                trailing_1_day_impressions=0,
-                trailing_7_day_opens=0,
-                trailing_7_day_impressions=0,
-                trailing_14_day_opens=0,
-                trailing_14_day_impressions=0,
-                trailing_28_day_opens=99,
-                trailing_28_day_impressions=999,
-                created_at=0,
-                expires_at=0
-            ),
-        }
+        metrics = generate_metrics_model_dict(
+            id='home/999',
+            trailing_28_day_opens=99,
+            trailing_28_day_impressions=999,
+        )
 
         sampled_recs = thompson_sampling(recs, metrics)
         # this needs to be a set since order isn't guaranteed in single trial
         assert {item.item_id for item in sampled_recs} == {"999", "333"}
 
     def test_invalid_prior(self):
         recs = generate_recommendations(['999'])
-        metrics = {
-            'default': MetricsModel(
-                id='home/default',
-                trailing_1_day_opens=0,
-                trailing_1_day_impressions=0,
-                trailing_7_day_opens=0,
-                trailing_7_day_impressions=0,
-                trailing_14_day_opens=0,
-                trailing_14_day_impressions=0,
-                trailing_28_day_opens=99,
-                trailing_28_day_impressions=-14,
-                created_at=0,
-                expires_at=0,
-            ),
-        }
+        metrics = generate_metrics_model_dict(
+            id = 'home/default',
+            trailing_28_day_opens=99,
+            trailing_28_day_impressions=-14,
+        )
 
         sampled_recs = thompson_sampling(recs, metrics)
         # this needs to be a set since order isn't guaranteed in single trial
         assert {item.item_id for item in sampled_recs} == {"999"}
 
-    # Moved from a previous thompson sampling test file
-    def test_rank_by_ctr_over_n_trials(self, ntrials=99):
+    def test_invalid_trailing_period(self):
+        """
+        An exception should be raised is the trailing period does not exist on any metrics model
+        :return:
+        """
+        # Invalid trailing_period
+        with pytest.raises(ValueError):
+            thompson_sampling(
+                generate_recommendations(['999']),
+                metrics=generate_metrics_model_dict(),
+                trailing_period=123  # Model does not have 123 day trailing metrics
+            )
+        # Invalid trailing_period for Firefox New Tab metrics
+        with pytest.raises(ValueError):
+            thompson_sampling(
+                generate_recommendations(['000-999']),
+                metrics=generate_firefox_metrics(['000-999']),
+                trailing_period=7  # MetricsModel has 7 day trailing period, but FirefoxNewTabMetricsModel does not.
+            )
+
+    @pytest.mark.parametrize("thompson_sampling_function,metrics", [
```

Comment:
I'm not usually a fan of parametrized tests, but this is _exactly_ the use case where I think they make sense 🙂